### DISCUSSION TITLE
Unspecified image with WCS

### DISCUSSION THREAD

**Discussion by swood-ecology** [eblondel/ows4R#127] [OGC WCS] [answered]
Hi, when I try to get coverage data from an FAO server I get either an internal error or an external one related to an unspecified image. 

Here's the setup code:

```
require(ows4)

# Create WCS Client
WCS <- WCSClient$new(
  url="https://io.apps.fao.org/geoserver/wcs", 
  serviceVersion="1.1.1", 
  logger="INFO"
)

# Get capabilities
caps <- WCS$getCapabilities()

# Get coverage summary
cov <- caps$findCoverageSummaryById("GLEAM3:a_prod")

# Describe a coverage
cov_des <- cov$getDescription()
```

Here's where I request the data:
```
# Get coverage data
cov_data <- cov$getCoverage()
```
This gives the following error:
```
> cov_data <- cov$getCoverage()
Error in `.rowNamesDF<-`(x, value = value) : invalid 'row.names' length
```
I figured this might be an issue of not having a bounding box, so I created a global one:

```
# Try with global bounding box
box <- OWSUtils$toBBOX(180,90,-180,-90)
cov_data <- cov$getCoverage(bbox=box)
```

This produced the following error:
```
> cov_data <- cov$getCoverage(bbox=box)
[ows4R][WARN] WCSCoverageSummary - Coverage without temporal dimension: 'time' argument is ignored 
[ows4R][WARN] WCSCoverageSummary - Coverage without vertical dimension: 'elevation' argument is ignored 
[ows4R][INFO] WCSGetCoverage - Fetching https://io.apps.fao.org/geoserver/wcs?service=WCS&version=1.1.1&identifier=GLEAM3:a_prod&boundingbox=-180,180,-90,90,urn:ogc:def:crs:EPSG::4326&format=image/tiff&request=GetCoverage 
Downloading: 300 B     [ows4R][ERROR] WCSCoverageSummary - Error while getting coverage: org.geotools.coverage.grid.InvalidGridGeometryException: Unspecified image&apos;s size.
Unspecified image&apos;s size. 
```

Does anyone have any insight into what I might be doing wrong?

**Comment by eblondel** [comment]
I'm looking at it; From what I see, this is not a problem with the `getCoverage` but rather with the decoding of the coverage description, for which some element (envelope) is not well handled.

**Reply by swood-ecology** [reply]
Does that mean it's something that I'm doing wrong in not knowing what to pass to the getCoverage function? Or is it something missing in the data from FAO?

**Reply by eblondel** [reply]
You don't do anything wrong; It could be an issue with the way Geoserver encodes the coverage description, this one in particular, or even some bug in ows4R :-)

**Reply by eblondel** [reply, marked as answer]
I could make it work using version `2.0.1` of the WCS service. One note when using version 2 the separator between the namespace and the layer name is not `:` but a double underscoe `__`. Layer name would be `GLEAM3__a_prod`.

I realized this server fails 1 out of 2 attempts to get coverage. Don't be surprised if the first request fails, in general the 2d works. I suspect they have a load balancer that delegates request to one service instance that is down behind, which would explain why 1 request out of 2 is failing.

See this code

```R
require(ows4R)

# Create WCS Client
WCS <- WCSClient$new(
  url="https://io.apps.fao.org/geoserver/wcs", 
  serviceVersion="2.0.1", 
  logger="INFO"
)

# Get capabilities
caps <- WCS$getCapabilities()

# Get coverage summary
cov <- caps$findCoverageSummaryById("GLEAM3__a_prod")

# Describe a coverage
cov_des <- cov$getDescription()

# Get coverage
cov_data <- cov$getCoverage()

# plot
require(terra)
plot(cov_data)
```
![image](https://github.com/eblondel/ows4R/assets/1324396/20b90193-21df-47a5-9585-9ac81909a363)
